### PR TITLE
add json_type function and compare json with primitive types.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -252,6 +248,13 @@ func (s *builtinCompareSig) eval(row []types.Datum) (d types.Datum, err error) {
 		return
 	}
 
+	if s.op != opcode.NullEQ {
+		if aa, bb, err := types.CoerceDatum(sc, a, b); err == nil {
```
Comment (hicqu) [inline expression/builtin_compare.go]: here, coerce error should be skiped, or `select * from t where text_field = 0.3` will maybe fail.
what do you think?

**Comment by shenli** [comment]
Please add unit tests for newly added code.

**Comment by hicqu** [comment]
address coments. PTAL, thanks.

File: util/types/json/compare.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	} else if x-y < 0 {
+		return -1
+	}
+	return 1
+}
+
+// jsonTypePrecedences is for compare two json.
+// See: https://dev.mysql.com/doc/refman/5.7/en/json.html#json-comparison
+var jsonTypePrecedences = map[string]int{
+	"BLOB":     -1,
```
Comment (shenli) [inline util/types/json/compare.go]: Seems there is no type named "BLOB" in json.Type().

File: util/types/json/compare.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		case jsonObject:
+			// only equal is defined on two json objects.
+			// larger and smaller are not defined.
+			s1 := Serialize(x)
+			s2 := Serialize(j2)
+			cmp = bytes.Compare(s1, s2)
+		default:
+			cmp = 0
+		}
+	} else if (precedence1 == -7 && precedence2 == -11) || (precedence1 == -11 && precedence2 == -7) {
```
Comment (shenli) [inline util/types/json/compare.go]: Add comment for this branch。

File: util/types/json/compare.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import (
+	"bytes"
+	"reflect"
+	"strings"
+)
+
+// compareFloat64 returns an integer comparing the float64 x to y,
+// allowing precision loss.
+func compareFloat64PrecisionLoss(x, y float64) int {
+	var EPSILON = 0.00000001
```
Comment (shenli) [inline util/types/json/compare.go]: Use a const and add comment.

File: util/types/json/compare.go (hunk 2 above)
Comment (hicqu) [inline util/types/json/compare.go]: Yes, it's just from MySQL 5.7 document. Maybe in future we will add those types into JSON, so I prefer keep it. what do you think?

File: expression/builtin_compare.go (hunk 1 above)
Comment (hanfei1991) [inline expression/builtin_compare.go]: When will this function report an error?

Comment (hicqu) [inline expression/builtin_compare.go]: for example, coerce 'hello' to decimal or float will report an error.

File: util/types/compare.go [hunk 5]
```diff
@@ -1,7 +1,3 @@
-// Copyright 2014 The ql Authors. All rights reserved.
-// Use of this source code is governed by a BSD-style
-// license that can be found in the LICENSES/QL-LICENSE file.
-
```
Comment (hicqu) [inline util/types/compare.go]: we don't need ql lisence anymore. see #3313

**Comment by hicqu** [comment]
all CI have passwd. PTAL again, thanks!

File: util/types/json/compare.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+				cmp, _ = CompareJSON(elem1, elem2)
+				if cmp != 0 {
+					return
+				}
+			}
+			cmp = len(x) - len(y)
+		case jsonObject:
+			// only equal is defined on two json objects.
+			// larger and smaller are not defined.
+			s1 := Serialize(x)
```
Comment (XuHuaiyu) [inline util/types/json/compare.go]: can we replace x with j1, may be more symmetric...

File: util/types/json/compare.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	return string(j.(jsonString))
+}
+
+// CompareJSON compares two json object.
+func CompareJSON(j1 JSON, j2 JSON) (cmp int, err error) {
+	precedence1 := jsonTypePrecedences[j1.Type()]
+	precedence2 := jsonTypePrecedences[j2.Type()]
+
+	if precedence1 == precedence2 {
+		if precedence1 == -12 {
```
Comment (XuHuaiyu) [inline util/types/json/compare.go]: use jsonTypePrecedences["NULL"] here may be more explicit.
the same as line 121 and line 125

File: util/types/json/impl.go [hunk 8]
```diff
@@ -41,7 +41,14 @@ func normalize(in interface{}) JSON {
 		return jsonLiteralFalse
 	case nil:
 		return jsonLiteralNil
+	case int64:
+		return jsonInt64(t)
 	case float64:
+		// json.Unmarshal converts all number to float64,
```
Comment (XuHuaiyu) [inline util/types/json/impl.go]: s/number/numbers

File: util/types/json/impl.go [hunk 9]
```diff
@@ -41,7 +41,14 @@ func normalize(in interface{}) JSON {
...
+		// so we need convert integer back.
```
Comment (XuHuaiyu) [inline util/types/json/impl.go]: s/need/need to

File: util/types/json/impl.go [hunk 10]
```diff
@@ -41,7 +41,14 @@ func normalize(in interface{}) JSON {
...
+		if float64(int64(t)) == t {
```
Comment (XuHuaiyu) [inline util/types/json/impl.go]: what if t == 1.0, we consider it as an integer?

File: util/types/json/compare.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+package json
+
+import (
+	"bytes"
+	"reflect"
+	"strings"
+)
+
+// floatEpsilon is for compare two float value allowing precision loss.
+const floatEpsilon float64 = 0.00000001
```
Comment (hanfei1991) [inline util/types/json/compare.go]: 1.e-8 is clearer?

File: util/types/json/json.go [hunk 12]
```diff
@@ -26,6 +26,13 @@ type JSON interface {
 	encode(buffer *bytes.Buffer)
 	// String returns JSON text.
 	String() string
+	// Type returns type string.
+	Type() string
+}
+
+// CreateJSON create a JSON from in. return nil if in is not valid JSON type.
```
Comment (XuHuaiyu) [inline util/types/json/json.go]: s/return/Return
s/valid/a valid

File: util/types/json/json_test.go [hunk 13]
```diff
@@ -74,3 +74,86 @@ func (s *testJSONSuite) TestParseFromString(c *C) {
 	var jstr2 = j1.String()
 	c.Assert(jstr2, Equals, `{"a":[1,"2",{"aa":"bb"},4,null],"b":true,"c":null}`)
 }
+
+func (s *testJSONSuite) TestJSONType(c *C) {
```
Comment (XuHuaiyu) [inline util/types/json/json_test.go]: what will ParseFromString(`1.0`) return?

File: util/types/json/impl.go [hunk 14]
```diff
@@ -57,8 +64,9 @@ func normalize(in interface{}) JSON {
 			array[i] = normalize(elem)
 		}
 		return jsonArray(array)
+	default:
+		panic("unsupported json type")
```
Comment (XuHuaiyu) [inline util/types/json/impl.go]: should we support [jsonDecimal](https://github.com/mysql/mysql-server/blob/5.7/sql/json_dom.cc#L2941)?

File: util/types/json/compare.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+	} else if (precedence1 == -7 && precedence2 == -11) || (precedence1 == -11 && precedence2 == -7) {
+		// tidb treat boolean as integer, but boolean is different from integer in JSON.
+		// so we need convert them to same type and then compare.
+		var x, y float64
+		if precedence1 == -7 {
+			x = float64(j1.(jsonLiteral))
+			y = jsonAsFloat64(j2)
+		} else {
+			x = jsonAsFloat64(j1)
+			y = float64(*j2.(*jsonLiteral))
```
Comment (hanfei1991) [inline util/types/json/compare.go]: Why line 130 and 125 are different.

File: util/types/json/compare.go [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+func jsonAsFloat64(j JSON) float64 {
+	if j.Type() == "INTEGER" {
+		if reflect.TypeOf(j).Kind() == reflect.Ptr {
+			return float64(*j.(*jsonInt64))
+		}
+		return float64(j.(jsonInt64))
+	} else if j.Type() == "DOUBLE" {
+		if reflect.TypeOf(j).Kind() == reflect.Ptr {
```
Comment (hanfei1991) [inline util/types/json/compare.go]: Why sometimes it is a pointer but sometimes not?

Comment (hicqu) [inline util/types/json/compare.go]: I don't know, either. could you help me for this?

File: util/types/json/impl.go (hunk 10 above)
Comment (hicqu) [inline util/types/json/impl.go]: yes, but I believe there is no solution for that except writing UnmarshalJSON by ourselves.
Any good suggestions?

File: util/types/json/impl.go (hunk 14 above)
Comment (hicqu) [inline util/types/json/impl.go]: after serialize, there is no jsonDecimal but jsonDouble instead.
you can see in MySQL 5.7 source code, Decimal, Integer and Double have same compare-martix.

File: util/types/json/json_test.go (hunk 13 above)
Comment (hicqu) [inline util/types/json/json_test.go]: If 1.0 is `"1.0"`, it will return a jsonDOUBLE.

**Comment by hicqu** [comment]
addressed comments. PTAL again, thanks.

**Comment by hicqu** [comment]
PTAL again, thanks!

File: util/types/json/compare.go (hunk 16 above)
Comment (shenli) [inline util/types/json/compare.go]: I am confused too.

File: util/types/json/impl.go (hunk 10 above)
Comment (shenli) [inline util/types/json/impl.go]: This maybe a better solution: https://stackoverflow.com/questions/22343083/json-marshaling-with-long-numbers-in-golang-gives-floating-point-number

File: expression/builtin_compare.go (hunk 1 above)
Comment (shenli) [inline expression/builtin_compare.go]: Any test case covers this?

File: util/types/json/compare.go (hunk 16 above)
Comment (shenli) [inline util/types/json/compare.go]: Could you show me a case?

File: expression/builtin_compare.go (hunk 1 above)
Comment (hicqu) [inline expression/builtin_compare.go]: now `select * from t where json_field = 0.3` is coverd, but `select * from t where text_field = 0.3` is not. that will be later, after we fix compare text with decimal.

File: util/types/json/json_test.go (hunk 13 above)
Comment (hicqu) [inline util/types/json/json_test.go]: Now, it returns jsonDOUBLE.

**Comment by hanfei1991** [comment]
LGTM

File: util/types/json/compare.go [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+package json
+
+import (
+	"bytes"
+	"reflect"
+	"strings"
+)
+
+// floatEpsilon is for compare two float value allowing precision loss.
```
Comment (XuHuaiyu) [inline util/types/json/compare.go]: I'm not sure will the folloing description be more explicit.

floatEpsilon is the acceptable error precision when comparing two floating numbers.

File: util/types/json/compare.go [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+const floatEpsilon float64 = 1.e-8
+
+// compareFloat64 returns an integer comparing the float64 x to y,
```
Comment (XuHuaiyu) [inline util/types/json/compare.go]: s/compareFloat64/compareFloat64PrecisionLoss
move this to util/types/compare.go ?

**Comment by shenli** [comment]
LGTM
/cc @AndreMouche

File: util/types/json/compare.go [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+	} else if x-y < 0 {
+		return -1
+	}
+	return 1
+}
+
+// jsonTypePrecedences is for comparing two json.
+// See: https://dev.mysql.com/doc/refman/5.7/en/json.html#json-comparison
+var jsonTypePrecedences = map[string]int{
+	"BLOB":     -1,
```
Comment (tiancaiamao) [inline util/types/json/compare.go]: why use negative number? convert them to uint would cause problem.

Comment (hicqu) [inline util/types/json/compare.go]: we will never convert them to uint. we just want a number list, order by desc. so I begin with -1, and so on -2, -3, ...

**Pull request merged**